runtime-modules/council/src/lib.rs

```diff
@@ -552,12 +557,35 @@ impl<T: Trait> ReferendumConnection<T> for Module<T> {
     }
 
     /// Check that it is a proper time to release stake.
-    fn can_release_vote_stake() -> Result<(), Error<T>> {
-        // ensure it's proper time to release stake
-        match Stage::<T>::get().stage {
-            CouncilStage::Idle => (),
-            _ => return Err(Error::CantReleaseStakeNow),
-        };
+    fn can_release_vote_stake(
+        vote: &CastVoteOf<T>,
+        current_voting_cycle_id: &u64,
```

Comment:
Why are you passing in a cycle id? This is the announcement period nr, right? This is stored in `AnnouncementPeriodNr`?

Reply:
This comment made me look at referendum to see where this id is coming from, and it is from a separate

```
        /// Index of the current referendum cycle. It is incremented everytime referendum ends.
        pub CurrentCycleId get(fn current_cycle_id) config(): u64;
```

This is only updated in `reset_referendum`, which is called only from `conclude_referendum`, which is called only from `end_reveal_period`, which at the same time informs the rest of the runtime about an ended election with `T::process_results(&winners);`. So I wonder, is there a need for having a separate representation of cycle id here? Its not clear to me, please respond.

If its not required, then suggest a fix in a separate issue, solve that in a separate PR, have me review it, when its accepted, reconcile this PR with that one by fixing the signature for this routine and anything else.

Reply:
`AnnouncementPeriodNr` is a number of announcement phases, while the referendum's `CurrentCycleId` is the number of elections held. These two will differ as there can be more announcement phases than elections. 

Your note on `reset_referendum`, etc., made me notice a bug there. `CurrentCycleId` is now increased when the referendum concludes as well as when it is restarted using `force_start`. This didn't create a problem for a council but was a mistake anyway. Fixed

Reply:
Ah, yes I agree, those are of course different, but then my question is: do we need to have the election id as separate? 

For the purpose of the referendum, we want to distinguish votes from different election rounds, hence couldn't the council just pass in the id to be used for the given election, and use it's announcement number as that id, and that way there is no need to maintain independent state in the referendum module from one cycle to the next? You can simply accept the unique id from the starting call of the caller module (election), and store it as part of the state of the current election, and then have it expire once the referendum goes to idle state. 

What do you think? Is this safer, or not? the caller can of course make a mistake and pass in non-unique values across calls also, this is currently not possible.

Reply:
It took me a while but I think I understood the bug you found, however the result now is that you are effectively having a dummy value for the cycle Id `Inactive` stage. So I think its basically like this:

1. You start in Voting stage of a new referendum with id = `N`
2. When reveal stage is over the new id is `N+1`, however this value is not ever written back anywehre, no vote is cast with this value, nor is there any referendum so to speak.
3. A new voting stage is started with id = `N+2`

If this is correct then I think its quite akward, and I also think part of the reason it took me so long to figure out whether things were correct in `can_unlock_vote_stake` this time around, in terms of all of the id compairson checks, is that we have a cycle id set when there is not cycle going. 

My proposal above would implictly solve this problem, as there is only an id for an election when there is an election, when one is in the `Inactive` stage, then there is no cycle id to even compare votes against, clearing things up.

Reply:
This is a really good question. After giving it some thought, I think it is a viable solution, but it would introduce the potential to start a referendum with the bad index, as you mentioned (breaking ascending order of indices). Also, this externalization of the election counter creates a need to reimplement such a counter in each module that uses referendum, thus reducing reusability and easy-to-use (in the future as we are using it only in the council right now).

On the other hand, I don't think there are any security or terrible performance issues with keeping the index; thus, I am rather for keeping the index present in the referendum.

Reply:
Replying to your second post: 

The id doesn't change between voting and revealing stages:
1. voting stage starts with id = `N`
2. revealing stage still has id = `N`
3. when a new voting stage starts, it has id = `N+1`

Reply:
> The id doesn't change between voting and revealing stages:

I know, that is not what I wrote, I wrote `When reveal stage is over the new`

|  Stage  |  Id  |
|---|---|
| Voting  |  `N`   |
| Revealing  |  `N`   |
| Idle  |  `N+1`  (dummy value) |
| Voting  |  `N+2`   |
| Revealing  |  `N+2`   |

> I don't think there are any security or terrible performance issues with keeping the index

I agree it's not about performance at all, it's about clarity. If we are going to keep this Id I don't know what semantics we give it that gives clarity to the reader. The fact that I even missed the latest fix you added in `897b63f` makes me think we do not have sufficient clarity.

> election counter creates a need to re-implement such a counter in each module that uses referendum

This is true, but remember that this voting system is inherently periodic, this is why we have the concept of votings having around from old elections while new elections get started. I think it will be very natural for any client module to itself reflect this periodicity in some way, as it has to provide the candidate pool and so on over time. Stated differently, I can't think of a user module of `referendum` which only ever invokes a single election.

Reply:
I think there was a dummy value as you described at one point, but it was later reworked, so `N` "is" increased at the referendum started, so there is no such dummy value.

Anyway, I removed the counter from the referendum and adapted the council.